src/main/java/org/opensearch/ad/rest/handler/ModelValidationActionHandler.java

```diff
@@ -0,0 +1,718 @@
+                        CommonErrorMessages.MODEL_VALIDATION_FAILED_UNEXPECTEDLY,
+                        DetectorValidationIssueType.MODEL_VALIDATION_ISSUE,
+                        ValidationAspect.MODEL
+                    )
+                );
+            return null;
+        }
+        Histogram aggregate = aggs.get("agg");
+        if (aggregate == null) {
+            listener.onFailure(new IllegalArgumentException("Failed to find valid aggregation result"));
+            return null;
+        }
+        return aggregate;
+    }
+
+    private void processRawDataResults(SearchResponse response, long latestTime) {
+        Histogram aggregate = checkBucketResultErrors(response);
+        if (aggregate == null) {
+            return;
+        }
+        double fullBucketRate = processBucketAggregationResults(aggregate);
+        if (fullBucketRate < CONFIG_BUCKET_MINIMUM_SUCCESS_RATE) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.RAW_DATA_TOO_SPARSE,
+                        DetectorValidationIssueType.INDICES,
+                        ValidationAspect.MODEL
+                    )
+                );
+        } else {
+            checkDataFilterSparsity(latestTime);
+        }
+    }
+
+    private void checkDataFilterSparsity(long latestTime) {
+        AggregationBuilder aggregation = getBucketAggregation(latestTime);
+        BoolQueryBuilder query = QueryBuilders.boolQuery().filter(anomalyDetector.getFilterQuery());
+        SearchSourceBuilder searchSourceBuilder = getSearchSourceBuilder(query, aggregation);
+        SearchRequest searchRequest = new SearchRequest(anomalyDetector.getIndices().toArray(new String[0])).source(searchSourceBuilder);
+        client.search(searchRequest, ActionListener.wrap(response -> processDataFilterResults(response, latestTime), listener::onFailure));
+    }
+
+    private void processDataFilterResults(SearchResponse response, long latestTime) {
+        Histogram aggregate = checkBucketResultErrors(response);
+        if (aggregate == null) {
+            return;
+        }
+        double fullBucketRate = processBucketAggregationResults(aggregate);
+        if (fullBucketRate < CONFIG_BUCKET_MINIMUM_SUCCESS_RATE) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.FILTER_QUERY_TOO_SPARSE,
+                        DetectorValidationIssueType.FILTER_QUERY,
+                        ValidationAspect.MODEL
+                    )
+                );
+        } else if (anomalyDetector.isMultientityDetector()) {
+            getTopEntityForCategoryField(latestTime);
+        } else {
+            try {
+                checkFeatureQuery(latestTime);
+            } catch (Exception ex) {
+                logger.error(ex);
+                listener.onFailure(ex);
+            }
```

Comment:
Can you add a comment about these `else` blocks? From my understanding it means that the filter query is ok, so now you are checking the features / category fields, in case the sparsity problem likes there